src/node_file.h

```diff
@@ -17,60 +17,48 @@ using v8::Value;
 
 namespace fs {
 
-class FSReqWrap: public ReqWrap<uv_fs_t> {
+enum class Ownership { COPY, MOVE };
+
+class FSReqWrap : public ReqWrap<uv_fs_t> {
  public:
-  enum Ownership { COPY, MOVE };
+  FSReqWrap(Environment* env, Local<Object> req)
+      : ReqWrap(env, req, AsyncWrap::PROVIDER_FSREQWRAP) {
+    Wrap(object(), this);
+  }
 
-  inline static FSReqWrap* New(Environment* env,
-                               Local<Object> req,
-                               const char* syscall,
-                               const char* data = nullptr,
-                               enum encoding encoding = UTF8,
-                               Ownership ownership = COPY);
+  virtual ~FSReqWrap() {
+    ReleaseEarly();
+    ClearWrap(object());
+  }
 
-  inline void Dispose();
+  virtual void Init(const char* syscall,
+                    const char* data = nullptr,
+                    enum encoding encoding = UTF8,
+                    Ownership ownership = Ownership::COPY);
 
+  virtual void FillStatsArray(const uv_stat_t* stat);
   virtual void Reject(Local<Value> reject);
   virtual void Resolve(Local<Value> value);
+  virtual void ResolveStat();
 
   void ReleaseEarly() {
-    if (data_ != inline_data()) {
-      delete[] data_;
-      data_ = nullptr;
+    if (buffer_ != nullptr) {
+      delete[] buffer_;
+      buffer_ = nullptr;
     }
   }
 
   const char* syscall() const { return syscall_; }
   const char* data() const { return data_; }
-  const enum encoding encoding_;
+  enum encoding encoding() const { return encoding_; }
 
   size_t self_size() const override { return sizeof(*this); }
 
- protected:
-  FSReqWrap(Environment* env,
-            Local<Object> req,
-            const char* syscall,
-            const char* data,
-            enum encoding encoding)
-      : ReqWrap(env, req, AsyncWrap::PROVIDER_FSREQWRAP),
-        encoding_(encoding),
-        syscall_(syscall),
-        data_(data) {
-    Wrap(object(), this);
-  }
-
-  virtual ~FSReqWrap() {
-    ReleaseEarly();
-    ClearWrap(object());
-  }
-
-  void* operator new(size_t size) = delete;
-  void* operator new(size_t size, char* storage) { return storage; }
-  char* inline_data() { return reinterpret_cast<char*>(this + 1); }
-
  private:
+  enum encoding encoding_ = UTF8;
   const char* syscall_;
-  const char* data_;
+  const char* data_ = nullptr;
+  const char* buffer_ = nullptr;
```

Comment:
`buffer_` and `ReleaseEarly()` are essentially unused now, as far as I can tell?

Reply:
`Ownership::COPY `(and therefore `buffer_`) is used within the `WriteString()` function, line 1008.
`ReleaseEarly()` is used with `FSReqAfterScope` (see line 157).

As far as I can tell, those are the only current uses and can likely be refactored away at some point but I think I'd rather prefer to avoid doing that in this PR

Reply:
oh wait... ha! yeah, you're right. hold on a sec...